File: app/src/main/java/com/rayworks/shadowwrapper/ViewUtils.java
package com.rayworks.shadowwrapper;

import android.content.res.Resources;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.support.annotation.ColorRes;
import android.support.annotation.DimenRes;
import android.support.v4.content.ContextCompat;
import android.view.Gravity;
import android.view.View;

public class ViewUtils {

    public static Drawable generateBackgroundWithShadow(
            View view,
            @ColorRes int backgroundColor,
            @DimenRes int cornerRadius,
            @ColorRes int shadowColor,
            int elevationValue,
            int shadowGravity) {
        Resources resources = view.getContext().getResources();

        float cornerRadiusValue = resources.getDimension(cornerRadius);
        int shadowColorValue = ContextCompat.getColor(view.getContext(), shadowColor);
        int backgroundColorValue = ContextCompat.getColor(view.getContext(), backgroundColor);

        float[] outerRadius = {
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue,
            cornerRadiusValue
        };

        Rect shapeDrawablePadding = new Rect();
        shapeDrawablePadding.left = elevationValue;
        shapeDrawablePadding.right = elevationValue;

        int DY;
        switch (shadowGravity) {
            case Gravity.CENTER:
                shapeDrawablePadding.top = elevationValue;
                shapeDrawablePadding.bottom = elevationValue;
                DY = 0;
                break;
            case Gravity.TOP:
                shapeDrawablePadding.top = elevationValue * 2;
                shapeDrawablePadding.bottom = elevationValue;
                DY = -1 * elevationValue / 3;
                break;
            default:
            case Gravity.BOTTOM:
                shapeDrawablePadding.top = elevationValue;
                shapeDrawablePadding.bottom = elevationValue * 2;
                DY = elevationValue / 3;
                break;
        }

        ShapeDrawable shapeDrawable = new ShapeDrawable();
        shapeDrawable.setPadding(shapeDrawablePadding);

        Paint paint = shapeDrawable.getPaint();
        paint.setColor(backgroundColorValue);
        paint.setShadowLayer(elevationValue /*/3*/, 0, DY, shadowColorValue);

        view.setLayerType(View.LAYER_TYPE_SOFTWARE, paint);

        shapeDrawable.setShape(new RoundRectShape(outerRadius, null, null));

        // merge it to LayerDrawable and set Inset to drawable for not seeing cutting shadow
        LayerDrawable drawable = new LayerDrawable(new Drawable[] {shapeDrawable});
        int vertical = elevationValue /* * 2*/;
        int horizontal = elevationValue;
        drawable.setLayerInset(0, horizontal, vertical, horizontal, vertical);

        return drawable;
    }
}

File: shadowlib/src/main/java/com/rayworks/shadowlib/ViewUtils.kt
package com.rayworks.shadowlib

import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.os.Build
import android.view.Gravity
import android.view.View

object ViewUtils {
    /***
     * Applies the shadow effect
     */
    fun View.applyShadow(
        backgroundColorValue: Int,
        cornerRadiusValue: Float,
        shadowColorValue: Int,
        elevationValue: Int,
        shadowGravity: Int,
        insetLeft: Int = elevationValue,
        insetTop: Int = elevationValue,
        insetRight: Int = elevationValue,
        insetBottom: Int = elevationValue,
        dx: Float = 0F,
        dy: Float = 0F
    ) {
        background = generateBackgroundWithShadow(
            this, backgroundColorValue, cornerRadiusValue, shadowColorValue, elevationValue,
            shadowGravity, insetLeft, insetTop, insetRight, insetBottom, dx, dy
        )
    }

    /***
     * Generates a shadow drawable
     */
    private fun generateBackgroundWithShadow(
        view: View,
        backgroundColorValue: Int,
        cornerRadiusValue: Float,
        shadowColorValue: Int,
        elevationValue: Int,
        shadowGravity: Int,
        insetLeft: Int = elevationValue,
        insetTop: Int = elevationValue,
        insetRight: Int = elevationValue,
        insetBottom: Int = elevationValue,
        dx: Float = 0F,
        dy: Float = 0F

    ): Drawable {
        val outerRadius = FloatArray(8) { cornerRadiusValue }

        val shapeDrawablePadding = Rect()
        shapeDrawablePadding.left = elevationValue
        shapeDrawablePadding.right = elevationValue

        when (shadowGravity) {
            Gravity.CENTER -> {
                shapeDrawablePadding.top = elevationValue
                shapeDrawablePadding.bottom = elevationValue
            }
            Gravity.TOP -> {
                shapeDrawablePadding.top = elevationValue * 2
                shapeDrawablePadding.bottom = elevationValue
            }
            Gravity.BOTTOM -> {
                shapeDrawablePadding.top = elevationValue
                shapeDrawablePadding.bottom = elevationValue * 2
            }
            else -> {
                shapeDrawablePadding.top = elevationValue
                shapeDrawablePadding.bottom = elevationValue * 2
            }
        }

        val shapeDrawable = ShapeDrawable()
        shapeDrawable.setPadding(shapeDrawablePadding)

        val paint = shapeDrawable.paint
        paint.color = backgroundColorValue
        paint.setShadowLayer(elevationValue.toFloat(), dx, dy, shadowColorValue)

        // https://developer.android.com/guide/topics/graphics/hardware-accel
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P)
            view.setLayerType(View.LAYER_TYPE_HARDWARE, paint)
        else
            view.setLayerType(View.LAYER_TYPE_SOFTWARE, paint)

        // NB: Given that the RoundRectShape will be set as the view's background, you can only change its
        // dimension via the host view
        shapeDrawable.shape = RoundRectShape(outerRadius, null, null)

        // merge it to LayerDrawable and set Inset to drawable for not seeing cutting shadow
        val drawable = LayerDrawable(arrayOf<Drawable>(shapeDrawable))
        drawable.setLayerInset(0, insetLeft, insetTop, insetRight, insetBottom)

        return drawable
    }
}
